### ISSUE TITLE
[feature request] More methods for PackedSequence

### ISSUE THREAD

**Issue by dmitriy-serdyuk** [pytorch/pytorch#8921]
As a person who works a lot with recurrent networks and sequences, I wish it was easier to work with `PackedSequence`. I frequently find myself packing/unpacking sequence to perform some simple operation such as
```
sequence, lengths = pad_packed_sequence(input, batch_first=True)
sequence = elementwise_function(sequence)
pack_padded_sequence(sequence, lengths, batch_first=True)
```
 This complicates code, introduces bugs (setting `batch_first` incorrectly), and makes me write two versions of code for tensors and sequences.

`PaddedSequence` can be extended to act more like `torch.Tensor`. More specifically, it can be extended with following methods (I tried to choose ones which make sense):

# Element-wise unary operations (wrt to the tensor)

- abs, acos, asin, atan, atan2, ceil, clamp, contiguous, cos, cosh,  exp, expm1, frac, log, log10, log1p, log2, mul, pow, reciprocal, neg, renorm, round, rsqrt, sigmoid, sign, sin, sinh, sqrt, tan, tanh, trunc 
- fill_, pin_memory

# Boolean functions

- is_contiguous, is_pinned

# Element-wise binary operations

Two sequences can be added, subtracted, compared, etc when they have same shape and lengths.

- add, div, eq, fmod, ge, gt, le, lt, ne, remainder, sub
- map_

# Autograd functions

- detach

**Comment by t-vi** [comment]
The binary operations are outright dangerous unless you implement erring out when feeding sequences of different length.
Personally, I use an apply method
```
def elementwise_applyi(fn, arg):
    return torch.nn.utils.rnn.PackedSequence(fn(arg.data, args[0].batch_sizes))

def elementwise_apply_multi(fn, *args):
    # DOES NOT CHECK compatibility of packed sequences!
    return torch.nn.utils.rnn.PackedSequence(fn(*[(arg.data if type(arg)==torch.nn.utils.rnn.PackedSequence else arg) for arg in args]), args[0].batch_sizes)
```
but that has been all I ever needed. Can be used with `torch.x` functions, user-defined or lambdas.

**Comment by dmitriy-serdyuk** [comment]
Checking that sequences are the same length doesn't seem very difficult.

In 4.0 `PackedSequence` received `cuda()`, `cpu()` and type conversion methods, in 5.0 `to()`. It would be logical to continue this direction of making it to behave like `Tensor`.

**Comment by shengpu1126** [comment]
The documentation seems to suggest that users should never invoke the `__init__` method of `PackedSequence`. I agree that the `elementwise_apply` functions are a way to make this work, but it would be nice if we make the interface of `PackedSequence` and `Tensor` consistent. 

https://github.com/pytorch/pytorch/blob/0656ef483d1cd9041ad465652d3c151a60a96800/torch/nn/utils/rnn.py#L18

**Comment by floydluo** [comment]
I have a question about using `torch.nn.utils.rrn.pack_padded_sequence( input, lengths, batch_first)`. When using this function, the input sequences should be sorted by length in a decreasing order, as well as the length. What if I want to keep the original order of sequences in the order? Can this function support the input which is not sorted by length in a decreasing order?

**Comment by denkorzh** [comment]
also some aggregating operations would be quite useful, e.g. `mean()` with respect to varying lengths of sequences et cetera

**Comment by RylanSchaeffer** [comment]
What's the status of this? Coming soon, I hope :D

**Comment by jannessm** [comment]
I would be very interested in concatenation of packedsequences with other packedsequences or tensors!
